### PR TITLE
Automatic repair

### PR DESCRIPTION
**Pull request by domsj**
and also deprecate separate rebalance process

### CODE REVIEW

File: ocaml/src/automatic_repair.ml [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                                 []
+                                 [ Osd.Update.set_string
+                                     Osd_keys.test_key
+                                     (Lazy.force Osd_access.large_value)
+                                     Checksum.Checksum.NoChecksum
+                                     false ])
+                >>= function
+                | Osd.Ok ->
+                   let open Osd_state in
+                   osd_state.write <- Unix.gettimeofday () :: osd_state.write;
```
Comment (toolslive) [inline ocaml/src/automatic_repair.ml]: `Osd_state.add_write osd_state`

File: ocaml/src/automatic_repair.ml [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+                alba_client # with_osd
+                            ~osd_id
+                            (fun osd ->
+                             osd # multi_get
+                                 Osd.Low
+                                 [ Slice.wrap_string Osd_keys.test_key; ])
+                >>= function
+                | [ Some _ ] ->
+                   let open Osd_state in
+                   osd_state.read <- Unix.gettimeofday () :: osd_state.read;
```
Comment (toolslive) [inline ocaml/src/automatic_repair.ml]: add_read

File: ocaml/src/automatic_repair.ml [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+                   Lwt.return ()
+                | _ -> Lwt.return ())
+               (fun exn ->
+                Lwt.return ())
+           end
+         else
+           Lwt.return ())
+     | (AnotherAlba _, _)
+     | (Available, _) -> Lwt.return ())
```
Comment (toolslive) [inline ocaml/src/automatic_repair.ml]: I'm starting to think ClaimInfo could use a `with_this_alba`

File: ocaml/src/maintenance.ml [hunk 4]
```diff
@@ -44,38 +37,87 @@ class client ?(retry_timeout = 60.)
...
+    method failure_detect_all_osds : unit Lwt.t =
+      Lwt_extra2.run_forever
+        "failure_detect_all_osds"
+        (fun () ->
+         if maintenance_config.Maintenance_config.enable_auto_repair
+         then
+           begin
+             alba_client # mgr_access # list_all_osds >>= fun (_, osds) ->
+
+             (* failure detecting already decommissioned osds isn't that usefull *)
```
Comment (toolslive) [inline ocaml/src/maintenance.ml]: useful

File: ocaml/src/maintenance.ml [hunk 5]
```diff
@@ -1376,21 +1428,13 @@ class client ?(retry_timeout = 60.)
                      ~namespace_id);
                "repair by policy",
                (fun () -> self # repair_by_policy_namespace ~namespace_id);
+               "rebalance",
+               (fun () -> self # rebalance_namespace
+                               ~make_first_reverse:(fun () -> get_random_string 32,
```
Comment (toolslive) [inline ocaml/src/maintenance.ml]: we'll fix this after this merges in.

File: ocaml/src/automatic_repair.ml (hunk 3 above)
Comment (domsj) [inline ocaml/src/automatic_repair.ml]: Hmm, come to think of it: I'm using the wrong all_osds as a starting point.
It's probably better if I only query for osds claimed by this backend to begin with, I'll make an update.

**Pull request merged**